rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::{bindings, c_types, CStr};
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::mem::transmute;
+use core::pin::Pin;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
+        let drv_data = Box::into_raw(drv_data) as *mut c_types::c_void;
+        Ok(drv_data) as KernelResult<_>
+    };
+    // TODO don't we need to pin this?
+    let ptr = match f() {
+        Ok(ptr) => ptr,
+        Err(e) => return e.to_kernel_errno(),
+    };
+    rust_helper_platform_set_drvdata(pdev, ptr);
+    0
+}
+
+fn new_of_device_id(compatible: &CStr<'static>) -> KernelResult<bindings::of_device_id> {
+    // TODO:
+    // - fail at build time if compatible CStr doesn't fit.
+    // - can we do this safely without transmute?
+    let mut buf = [0_u8; 128];
+    if compatible.len() > buf.len() {
+        return Err(Error::EINVAL);
+    }
+    // PANIC: this will never panic: `compatible` is not longer than `buf`.
+    buf[..compatible.len()].copy_from_slice(compatible.as_bytes());
+    Ok(bindings::of_device_id {
+        // SAFETY: re-interpretation from [u8] to [i8] of same length is always safe.
+        compatible: unsafe { transmute::<[u8; 128], [i8; 128]>(buf) },
+        ..Default::default()
+    })
+}
+
+unsafe extern "C" fn remove_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let ptr = rust_helper_platform_get_drvdata(pdev);
+    let drv_data: Box<T::DrvData> = Box::from_raw(ptr as _);
```

Comment:
This needs a safety comment.
1. for the lifetime of `ptr` and reasoning that nothing else `free`s it
2. `Box::try_new()` needs some aditonal invariants see [Box::from_raw](https://doc.rust-lang.org/stable/std/boxed/struct.Box.html#safety-2) -> [box memory layout](https://doc.rust-lang.org/stable/std/boxed/index.html#memory-layout). In this case its probably sufficient to write something along the lines of "we allocated this pointer using `Box::try_new` and `Box::into_raw` so it is safe to turn back into a `Box`"

It should probably also be
```rust
let drv_data: Box<_> = Box::from_raw(ptr as T::DrvData);
```
instead as afaik it is usually prefered to have things influencing the safety of a call as close to the call itself as possible.